### ISSUE TITLE
Unable to use union return type in relation definition

### ISSUE THREAD

**Issue by Maxoulak** [nunomaduro/larastan#1181]
- Larastan Version: 2.1.2
- `--level` used: 5

### Description

Using union "|" return type on relations causes internal error.

### Laravel code where the issue was found

```php
// In Model App\Models\User
public function restaurants(): \Illuminate\Database\Eloquent\Relations\HasMany|\Illuminate\Database\Eloquent\Relations\BelongsToMany
    {
        return $this->role === RolesEnum::OWNER
            ? $this->hasMany(Restaurant::class, 'owner_id', 'id')
            : $this
                ->belongsToMany(Restaurant::class)
                ->withPivot('owner')
                ->withTimestamps();
    }
```

The above code causes the following error.
It has been solved when replacing :
```php
: \Illuminate\Database\Eloquent\Relations\HasMany|\Illuminate\Database\Eloquent\Relations\BelongsToMany
```
by
```php
: \Illuminate\Database\Eloquent\Relations\Relation
```

Stack Trace
```php
Internal error: Internal error: Call to undefined method PHPStan\Type\UnionType::getClassName() in file
     /var/www/html/app/Http/Controllers/RestaurantController.php

     Post the following stack trace to https://github.com/phpstan/phpstan/issues/new?template=Bug_report.md:
     #0 phar:///var/www/html/vendor/phpstan/phpstan/phpstan.phar/src/Reflection/ClassReflection.php(452):
     NunoMaduro\Larastan\Properties\ModelRelationsExtension->getProperty()
     #1 phar:///var/www/html/vendor/phpstan/phpstan/phpstan.phar/src/Type/ObjectType.php(150):
     PHPStan\Reflection\ClassReflection->getProperty()
     #2 phar:///var/www/html/vendor/phpstan/phpstan/phpstan.phar/src/Type/ObjectType.php(127):
     PHPStan\Type\ObjectType->getUnresolvedPropertyPrototype()
     #3 phar:///var/www/html/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/MutatingScope.php(3998):
     PHPStan\Type\ObjectType->getProperty()
     #4 phar:///var/www/html/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/MutatingScope.php(4005):
     PHPStan\Analyser\MutatingScope->getPropertyReflection()
     #5 phar:///var/www/html/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/MutatingScope.php(1933):
     PHPStan\Analyser\MutatingScope->propertyFetchType()
     #6 phar:///var/www/html/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/MutatingScope.php(1939):
     PHPStan\Analyser\MutatingScope->PHPStan\Analyser\{closure}()
     #7 phar:///var/www/html/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/MutatingScope.php(543):
     PHPStan\Analyser\MutatingScope->resolveType()
     #8 phar:///var/www/html/vendor/phpstan/phpstan/phpstan.phar/src/Reflection/ParametersAcceptorSelector.php(95):
     PHPStan\Analyser\MutatingScope->getType()
     #9 phar:///var/www/html/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/MutatingScope.php(3977):
     PHPStan\Reflection\ParametersAcceptorSelector::selectFromArgs()
     #10 phar:///var/www/html/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/MutatingScope.php(1886):
     PHPStan\Analyser\MutatingScope->methodCallReturnType()
     #11 phar:///var/www/html/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/MutatingScope.php(1892):
     PHPStan\Analyser\MutatingScope->PHPStan\Analyser\{closure}()
     #12 phar:///var/www/html/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/MutatingScope.php(543):
     PHPStan\Analyser\MutatingScope->resolveType()
     #13 phar:///var/www/html/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/MutatingScope.php(1886):
     PHPStan\Analyser\MutatingScope->getType()
     #14 phar:///var/www/html/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/MutatingScope.php(1892):
     PHPStan\Analyser\MutatingScope->PHPStan\Analyser\{closure}()
     #15 phar:///var/www/html/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/MutatingScope.php(543):
     PHPStan\Analyser\MutatingScope->resolveType()
     #16 phar:///var/www/html/vendor/phpstan/phpstan/phpstan.phar/src/Rules/FunctionReturnTypeCheck.php(52):
     PHPStan\Analyser\MutatingScope->getType()
     #17 phar:///var/www/html/vendor/phpstan/phpstan/phpstan.phar/src/Rules/Methods/ReturnTypeRule.php(43):
     PHPStan\Rules\FunctionReturnTypeCheck->checkReturnType()
     #18 phar:///var/www/html/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/FileAnalyser.php(102):
     PHPStan\Rules\Methods\ReturnTypeRule->processNode()
     #19 phar:///var/www/html/vendor/phpstan/phpstan/phpstan.phar/src/Node/ClassStatementsGatherer.php(95):
     PHPStan\Analyser\FileAnalyser->PHPStan\Analyser\{closure}()
     #20 phar:///var/www/html/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/NodeScopeResolver.php(451):
     PHPStan\Node\ClassStatementsGatherer->__invoke()
     #21 phar:///var/www/html/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/NodeScopeResolver.php(359):
     PHPStan\Analyser\NodeScopeResolver::PHPStan\Analyser\{closure}()
     #22 phar:///var/www/html/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/NodeScopeResolver.php(304):
     PHPStan\Analyser\NodeScopeResolver->processStmtNode()
     #23 phar:///var/www/html/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/NodeScopeResolver.php(466):
     PHPStan\Analyser\NodeScopeResolver->processStmtNodes()
     #24 phar:///var/www/html/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/NodeScopeResolver.php(304):
     PHPStan\Analyser\NodeScopeResolver->processStmtNode()
     #25 phar:///var/www/html/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/NodeScopeResolver.php(550):
     PHPStan\Analyser\NodeScopeResolver->processStmtNodes()
     #26 phar:///var/www/html/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/NodeScopeResolver.php(304):
     PHPStan\Analyser\NodeScopeResolver->processStmtNode()
     #27 phar:///var/www/html/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/NodeScopeResolver.php(516):
     PHPStan\Analyser\NodeScopeResolver->processStmtNodes()
     #28 phar:///var/www/html/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/NodeScopeResolver.php(275):
     PHPStan\Analyser\NodeScopeResolver->processStmtNode()
     #29 phar:///var/www/html/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/FileAnalyser.php(189):
     PHPStan\Analyser\NodeScopeResolver->processNodes()
     #30 phar:///var/www/html/vendor/phpstan/phpstan/phpstan.phar/src/Command/WorkerCommand.php(143):
     PHPStan\Analyser\FileAnalyser->analyseFile()
     #31
     phar:///var/www/html/vendor/phpstan/phpstan/phpstan.phar/vendor/evenement/evenement/src/Evenement/EventEmitterTrait.php(97):
     PHPStan\Command\WorkerCommand->PHPStan\Command\{closure}()
     #32 phar:///var/www/html/vendor/phpstan/phpstan/phpstan.phar/vendor/clue/ndjson-react/src/Decoder.php(110):
     _PHPStan_ae8980142\Evenement\EventEmitter->emit()
     #33
     phar:///var/www/html/vendor/phpstan/phpstan/phpstan.phar/vendor/evenement/evenement/src/Evenement/EventEmitterTrait.php(97):
     _PHPStan_ae8980142\Clue\React\NDJson\Decoder->handleData()
     #34 phar:///var/www/html/vendor/phpstan/phpstan/phpstan.phar/vendor/react/stream/src/Util.php(62):
     _PHPStan_ae8980142\Evenement\EventEmitter->emit()
     #35
     phar:///var/www/html/vendor/phpstan/phpstan/phpstan.phar/vendor/evenement/evenement/src/Evenement/EventEmitterTrait.php(97):
     _PHPStan_ae8980142\React\Stream\Util::_PHPStan_ae8980142\React\Stream\{closure}()
     #36 phar:///var/www/html/vendor/phpstan/phpstan/phpstan.phar/vendor/react/stream/src/DuplexResourceStream.php(154):
     _PHPStan_ae8980142\Evenement\EventEmitter->emit()
     #37 phar:///var/www/html/vendor/phpstan/phpstan/phpstan.phar/vendor/react/event-loop/src/StreamSelectLoop.php(200):
     _PHPStan_ae8980142\React\Stream\DuplexResourceStream->handleData()
     #38 phar:///var/www/html/vendor/phpstan/phpstan/phpstan.phar/vendor/react/event-loop/src/StreamSelectLoop.php(172):
     _PHPStan_ae8980142\React\EventLoop\StreamSelectLoop->waitForStreamActivity()
     #39 phar:///var/www/html/vendor/phpstan/phpstan/phpstan.phar/src/Command/WorkerCommand.php(106):
     _PHPStan_ae8980142\React\EventLoop\StreamSelectLoop->run()
     #40 phar:///var/www/html/vendor/phpstan/phpstan/phpstan.phar/vendor/symfony/console/Command/Command.php(259):
     PHPStan\Command\WorkerCommand->execute()
     #41 phar:///var/www/html/vendor/phpstan/phpstan/phpstan.phar/vendor/symfony/console/Application.php(856):
     _PHPStan_ae8980142\Symfony\Component\Console\Command\Command->run()
     #42 phar:///var/www/html/vendor/phpstan/phpstan/phpstan.phar/vendor/symfony/console/Application.php(259):
     _PHPStan_ae8980142\Symfony\Component\Console\Application->doRunCommand()
     #43 phar:///var/www/html/vendor/phpstan/phpstan/phpstan.phar/vendor/symfony/console/Application.php(157):
     _PHPStan_ae8980142\Symfony\Component\Console\Application->doRun()
     #44 phar:///var/www/html/vendor/phpstan/phpstan/phpstan.phar/bin/phpstan(94):
     _PHPStan_ae8980142\Symfony\Component\Console\Application->run()
     #45 phar:///var/www/html/vendor/phpstan/phpstan/phpstan.phar/bin/phpstan(95): _PHPStan_ae8980142\{closure}()
     #46 /var/www/html/vendor/phpstan/phpstan/phpstan(8): require('...')
     #47 /var/www/html/vendor/bin/phpstan(117): include('...')
     #48 {main}
     Child process error (exit code 1):
```

**Comment by szepeviktor** [comment]
[off] You could rewrite the code 👇 to be more readable.

```php
        return $this->role === RolesEnum::OWNER
            ? $this->hasMany(Restaurant::class, 'owner_id', 'id')
            : $this
                ->belongsToMany(Restaurant::class)
                ->withPivot('owner')
                ->withTimestamps();
```

**Comment by canvural** [comment]
This is fixed with #1187 But as a side note, if you use union types like this in your relation methods, to get better analysis by Larastan you'd need to add a return type PHPDoc that looks like this: 
`/** @return HasMany<Restaurant>|BelongsToMany<Restaurant> */`

Normally Larastan can do this automagically, but in this case because of union types are used, it can't.
